### DISCUSSION TITLE
How does RAUC select the target slot when installing?

### DISCUSSION THREAD

**Discussion by ghost** [rauc/rauc#1003] [Q&A]
So when having the usual A/B layout, 'rauc install' will install to the slot that is not booted.
But what if there is more then one slot to choose from? e.g. A/B/C for whatever reason.
Or in my case: a primary OS with A/B and a secondary OS also with A/B.
If I update the secondary OS while the primary OS is booted, rauc has 2 slots to choose from. How does it determine which one to use? Can I influence it's decision?

I know I can use write-slot to explicitly select the slot but I would prefer to have handlers and hooks available.

**Comment by JSydll** [comment]
From the [source code](https://github.com/rauc/rauc/blob/master/src/install.c#L292) it looks like it will use the first suitable inactive slot it finds iterating through all of the slots defined in the `system.conf`. I guess, this is deterministic as the `system.conf` does not change during runtime. So unless you patch the sources there may not be that many options to manipulate this behavior during runtime.

On a side note: The "whatever reason" in your example would be interesting ;) it might not be exactly the use case `rauc` was built for...

**Reply by jluebbe** [reply]
While RAUC supports more than 2 slots per class, this setup is not something we've deployed in any project so far, because we didn't have any use-case where that would have been useful. Accordingly there's no code to let RAUC choose between multiple inactive slots, as it's not clear what the criteria would be. Also note that slots in the same class should be equivalent from the perspective of the whole system, so having two different OSs in the same class wouldn't really work.

As handling the selection of target slots is one of RAUC key responsibility, I'd be interested to understand your requirements leading to having an A/B/C setup. For the primary/secondary OS case, where does the decision on which to boot happen?
